Update user's participant options

alxdarksage opened the pull request:
Researcher API to update a user's participant options. Can send some or all options in the JSON for this endpoint. Includes capability of endpoint to update user's dataGroups, and may replace it.

It may be possible to remove DataGroups object and the dataGroups API for researchers, as this is a superset of that functionality.

Other PRs have to come first before this PR is merged.

alxdarksage committed:
Adding API so researchers can update a user's participant options (some, none or all).

alxdarksage committed:
Move email out of path URI component, Heroku logs this

alxdarksage committed:
Throw 404 exception if there is no account for an email address.

DwayneJengSage commented on app/org/sagebionetworks/bridge/services/ParticipantService.java:
```diff
@@ -137,6 +142,17 @@ public StudyParticipant2 getParticipant(Study study, String email) {
+        checkNotNull(options);
+        
+        Account account = accountDao.getAccount(study, email);
+        String healthCode = getHealthCode(account);
```
You need a null-check between account and healthCode.

alxdarksage replied:
Added

DwayneJengSage commented on app/org/sagebionetworks/bridge/models/accounts/ParticipantOptions.java:
```diff
@@ -0,0 +1,61 @@
+    private final @Nullable Boolean notifyByEmail;
+    private final @Nullable String externalId;
+    private final @Nullable Set<String> dataGroups;
+    private final @Nullable LinkedHashSet<String> languages;
```
As is, if we add a new Participant Option, we'll need to update this class, the DAO, and every test that uses the ParticipantOptions constructor. This means that adding a new Participant Option is quadratically difficult with a rather large coefficient.

I feel like instead of top-level attributes, we should just have a string-to-string map. This way, we only need to update the enum and the direct consumer of the option.

alxdarksage replied:
We would still need to know for specific properties, if they are coming in as booleans or arrays, and deserialize them correctly into strings. This seems like the easiest way to do it since all the work is left to Jackson.

alxdarksage replied:
We discussed, we can not use Jackson and add a deserializer to each enum type, to convert these into a Map{String,String}, skipping entries if property doesn't exist. That can be saved directly in the corresponding DDB table.

DwayneJengSage replied:
Discussed this offline with Alx. The main concern here is that we have a quadratically growing list of things to change when we update ParticipantOptions, which makes this a maintainability nightmare.

On the flip side, we need to be able to parse the incoming request based on specific options and types. I think the best way to handle this would be to have the enum itself contain the parsing logic. This way, it's easy to maintain, type-safe, and the ParticipantOptions.java could still just be a map.

alxdarksage committed:
Switching to use of a map for values

alxdarksage committed:
- updating to main
- switching from bean to map of participant options

Conflicts:
	conf/routes
	test/org/sagebionetworks/bridge/play/controllers/ParticipantControllerTest.java

alxdarksage committed:
Remove ParticipantOptions

alxdarksage committed:
Removed unused cast

alxdarksage committed:
whitespace

alxdarksage committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/BridgePF into update-participant-data

Conflicts:
	app/org/sagebionetworks/bridge/play/controllers/ParticipantController.java
	app/org/sagebionetworks/bridge/services/ParticipantService.java
	conf/routes
	test/org/sagebionetworks/bridge/play/controllers/ParticipantControllerTest.java
	test/org/sagebionetworks/bridge/services/ParticipantServiceTest.java

DwayneJengSage commented on app/org/sagebionetworks/bridge/dao/ParticipantOption.java:
```diff
@@ -1,22 +1,74 @@
+    SHARING_SCOPE(SharingScope.NO_SHARING.name(), "sharingScope") {
+        public String deserialize(JsonNode node) {
+            checkNotNull(node);
+            return SharingScope.valueOf(node.asText().toUpperCase()).name();
```
This is user input. If the caller passes in an invalid sharing scope, Bridge server will throw a 500. We should add a try-catch here or in ParticipantController to wrap the exception as a BadRequest.

DwayneJengSage commented on app/org/sagebionetworks/bridge/dao/ParticipantOption.java:
```diff
@@ -1,22 +1,74 @@
+    
+    private static String arrayNodeToOrderedString(JsonNode node) {
+        Set<String> results = new LinkedHashSet<>();
+        ArrayNode array = (ArrayNode)node;
```
Similarly here, if the caller doesn't pass in an ArrayNode, Bridge server will throw a 500.

DwayneJengSage commented:
Looks mostly good. Just a few comments w.r.t user inputs and robustness.

alxdarksage committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/BridgePF into update-participant-data

alxdarksage committed:
Adding checks for JSON type errors as well as incorrect enum names (all throw BadRequestException).

alxdarksage commented:
Added checks for JSON type errors as well as incorrect SharingScope enum names. All throw BadRequestException.

alxdarksage committed:
Reorg for clarity

alxdarksage committed:
Check for null array member

DwayneJengSage commented:
Looks good. Ready to merge?
